grpc/metrics/src/main/java/io/helidon/grpc/metrics/GrpcMetrics.java

```diff
@@ -306,4 +370,132 @@ public void close(Status status, Metadata responseHeaders) {
             getMetric().update(1);
         }
     }
+
+
+    /**
+     * Implemented by classes that can create a metric name.
+     */
+    @FunctionalInterface
+    public interface NamingFunction {
+        /**
+         * Create a metric name.
+         *
+         * @param service    the service descriptor
+         * @param methodName the method name
+         * @param metricType the metric type
+         * @return the metric name
+         */
+        String createName(ServiceDescriptor service, String methodName, MetricType metricType);
+    }
+
+
+    /**
+     * An immutable holder of metrics information.
+     * <p>
+     * Calls made to mutating methods return a new instance
+     * of {@link MetricsRules} with the mutation applied.
+     */
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    static class MetricsRules {
+        /**
+         * The metric type.
+         */
+        private MetricType type;
+
+        /**
+         * The tags of the metric.
+         *
+         * @see org.eclipse.microprofile.metrics.Metadata
+         */
+        private Optional<HashMap<String, String>> tags = Optional.empty();
+
+        /**
+         * The description of the metric.
+         *
+         * @see org.eclipse.microprofile.metrics.Metadata
+         */
+        private Optional<String> description = Optional.empty();
+
+        /**
+         * The unit of the metric.
+         *
+         * @see org.eclipse.microprofile.metrics.Metadata
+         * @see org.eclipse.microprofile.metrics.MetricUnits
+         */
+        private Optional<String> units = Optional.empty();
+
+        /**
+         * The function to use to obtain the metric name.
+         */
+        private Optional<NamingFunction> nameFunction = Optional.empty();
+
+        private MetricsRules(MetricType type) {
+            this.type = type;
+        }
+
+        private MetricsRules(MetricsRules copy) {
+            this.type = copy.type;
+            this.tags = copy.tags;
+            this.description = copy.description;
+            this.units = copy.units;
+            this.nameFunction = copy.nameFunction;
+        }
+
+        /**
+         * Obtain the metric type.
+         *
+         * @return the metric type
+         */
+        MetricType type() {
+            return type;
+        }
+
+        /**
+         * Obtain the metrics metadata.
+         *
+         * @param service the service descriptor
+         * @param method the method name
+         * @return  the metrics metadata
+         */
+        org.eclipse.microprofile.metrics.Metadata metadata(ServiceDescriptor service, String method) {
+            String name;
+            if (nameFunction.isPresent()) {
+                name = nameFunction.get().createName(service, method, type);
+            } else {
+                name = (service.name() + "." + method).replaceAll("/", ".");
```

Comment:
I thought we said we were going to add `grpc.` prefix and `.<type>` suffix to all metric names?

As an aside, why not implement that as a default name function and turn the above condition into
```
String name = nameFunction.getOrDefault(DEFAULT_NAME_FUNCTION).createName(service, method, type);
```